File: app/src/main/java/com/bearcave/passageplanning/waypoints/database/WaypointsTable.java
package com.bearcave.passageplanning.waypoints.database;


import android.content.ContentValues;
import android.database.Cursor;

import com.bearcave.passageplanning.base.database.BaseTable;
import com.bearcave.passageplanning.data.database.ManagerListener;
import com.bearcave.passageplanning.tides.web.configurationitems.Gauge;

import java.util.LinkedHashMap;


public class WaypointsTable extends BaseTable<Waypoint> {

    public WaypointsTable(ManagerListener manager) {
        super(manager);
    }

    @Override
    public String getTableName() {
        return "waypoints";
    }

    @Override
    protected LinkedHashMap<String, String> createKeyToValueTypeHolder() {
        LinkedHashMap<String, String> typeHolder = new LinkedHashMap<>();

        typeHolder.put(KEY_ID,              INTEGER+PRIMARY_KEY+AUTOINCREMENT);
        typeHolder.put(KEY_NAME,            TEXT+NOT_NULL+UNIQUE);
        typeHolder.put(KEY_NOTE,            TEXT);
        typeHolder.put(KEY_CHARACTERISTIC,  TEXT);
        typeHolder.put(KEY_UKC,             FLOAT+NOT_NULL);
        typeHolder.put(KEY_LATITUDE,        DOUBLE+NOT_NULL);
        typeHolder.put(KEY_LONGITUDE,       DOUBLE+NOT_NULL);
        typeHolder.put(KEY_GAUGE,           INTEGER+NOT_NULL);

        return typeHolder;
    }


    @Override
    protected ContentValues getContentValue(Waypoint waypoint) {
        ContentValues values = new ContentValues();
        values.put(KEY_NAME, waypoint.getName());
        values.put(KEY_NOTE, waypoint.getNote());
        values.put(KEY_CHARACTERISTIC, waypoint.getCharacteristic());
        values.put(KEY_UKC, waypoint.getUkc());
        values.put(KEY_LATITUDE, waypoint.getLatitude());
        values.put(KEY_LONGITUDE, waypoint.getLongitude());
        values.put(KEY_GAUGE, waypoint.getGauge().getId());
        return values;
    }

    @Override
    protected Waypoint loadFrom(Cursor cursor) {
        int i = 0;
        return new Waypoint(
                cursor.getInt(i++),               // id
                cursor.getString(i++),            // name
                cursor.getString(i++),            // note
                cursor.getString(i++),            // characteristic
                cursor.getFloat(i++),             // ukc
                cursor.getDouble(i++),            // latitude
                cursor.getDouble(i++),            // longitude
                Gauge.getById(cursor.getInt(i))   // gauge
        );
    }

    private static final String KEY_ID = "id";
    private static final String KEY_NAME = "name";
    private static final String KEY_CHARACTERISTIC = "characteristic";
    private static final String KEY_NOTE = "note";
    private static final String KEY_UKC = "ukc";
    private static final String KEY_LONGITUDE = "longitude";
    private static final String KEY_LATITUDE = "latitude";
    private static final String KEY_GAUGE = "gauge_id";
}

File: app/src/main/java/com/bearcave/passageplanning/waypoints/database/WaypointsTable.kt
package com.bearcave.passageplanning.waypoints.database


import android.content.ContentValues
import android.database.Cursor
import com.bearcave.passageplanning.base.database.BaseTable
import com.bearcave.passageplanning.data.database.ManagerListener
import com.bearcave.passageplanning.tides.utils.Gauge
import java.util.*


class WaypointsTable(manager: ManagerListener) : BaseTable<Waypoint>(manager) {

    override val tableName: String
        get() = "waypoints"

    override fun createKeyToValueTypeHolder(): LinkedHashMap<String, String> {
        val typeHolder = LinkedHashMap<String, String>()

        typeHolder.put(KEY_ID, INTEGER + PRIMARY_KEY + AUTOINCREMENT)
        typeHolder.put(KEY_NAME, TEXT + NOT_NULL + UNIQUE)
        typeHolder.put(KEY_NOTE, TEXT)
        typeHolder.put(KEY_CHARACTERISTIC, TEXT)
        typeHolder.put(KEY_UKC, FLOAT + NOT_NULL)
        typeHolder.put(KEY_LATITUDE, DOUBLE + NOT_NULL)
        typeHolder.put(KEY_LONGITUDE, DOUBLE + NOT_NULL)
        typeHolder.put(KEY_GAUGE, INTEGER + NOT_NULL)

        return typeHolder
    }


    override fun getContentValue(element: Waypoint): ContentValues {
        val values = ContentValues()
        values.put(KEY_NAME, element.name)
        values.put(KEY_NOTE, element.note)
        values.put(KEY_CHARACTERISTIC, element.characteristic)
        values.put(KEY_UKC, element.ukc)
        values.put(KEY_LATITUDE, element.latitude)
        values.put(KEY_LONGITUDE, element.longitude)
        values.put(KEY_GAUGE, element.gauge.id)
        return values
    }

    override fun loadFrom(cursor: Cursor): Waypoint {
        var i = 0
        return Waypoint(
                cursor.getInt(i++), // id
                cursor.getString(i++), // humanCode
                cursor.getString(i++), // note
                cursor.getString(i++), // characteristic
                cursor.getFloat(i++), // ukc
                cursor.getDouble(i++), // latitude
                cursor.getDouble(i++), // longitude
                Gauge.getById(cursor.getInt(i))   // gauge
        )
    }

    companion object {

        val KEY_ID = "id"
        val KEY_NAME = "humanCode"
        val KEY_CHARACTERISTIC = "characteristic"
        val KEY_NOTE = "note"
        val KEY_UKC = "ukc"
        val KEY_LONGITUDE = "longitude"
        val KEY_LATITUDE = "latitude"
        val KEY_GAUGE = "gauge_id"
    }
}
